src/utils/hooks/useGetResponsesAndElementState.tsx

```diff
@@ -0,0 +1,127 @@
+import { useState, useEffect } from 'react'
+import { useGetElementsAndResponsesQuery, GetElementsAndResponsesQuery } from '../generated/graphql'
+import {
+  ResponsesByCode,
+  ResponsesFullByCode,
+  TemplateElementState,
+  EvaluatedElement,
+} from '../types'
+import evaluateExpression from '@openmsupply/expression-evaluator'
+
+const useGetResponsesAndElementState = (props: { serialNumber: string }) => {
+  const { serialNumber } = props
+  const [responsesByCode, setResponsesByCode] = useState({})
+  const [responsesFullByCode, setResponsesFullByCode] = useState({})
+  const [elementsExpressions, setElementsExpressions] = useState([{}])
+  const [elementsState, setElementsState] = useState({})
+  const [error, setError] = useState('')
+  const [loading, setLoading] = useState(true)
+  const { data, loading: apolloLoading, error: apolloError } = useGetElementsAndResponsesQuery({
+    variables: { serial: serialNumber },
+  })
+
+  useEffect(() => {
+    const error = checkForApplicationErrors(data)
+
+    if (error) {
+      setError(error)
+      setLoading(false)
+      return
+    }
+
+    if (apolloError) return
+
+    const applicationResponses = data?.applicationBySerial?.applicationResponses?.nodes
+
+    const templateSections = data?.applicationBySerial?.template?.templateSections?.nodes
+
+    const templateElements: TemplateElementState[] = []
+
+    templateSections?.forEach((sectionNode) => {
+      sectionNode?.templateElementsBySectionId?.nodes.forEach((element) => {
+        if (element) templateElements.push(element as TemplateElementState) // No idea why ...[spread] doesn't work here.
+      })
+    })
+
+    const currentResponses = {} as ResponsesByCode
+    const currentFullResponses = {} as ResponsesFullByCode
+
+    if (applicationResponses) {
+      applicationResponses.forEach((response) => {
+        const code = response?.templateElement?.code
+        if (code) {
+          currentResponses[code] = response?.value?.text
+          currentFullResponses[code] = response?.value
+        }
+      })
+    }
```

Comment:
```suggestion
    applicationResponses.forEach((response) => {
      const code = response.templateElement?.code
      if (code) {
        currentResponses[code] = response.value?.text
        currentFullResponses[code] = response.value
      }
    })
```

Since you already casted to `ApplicationResponse[]` at the top (and the checks will make sure you have received a valid object response).

Reply:
Thanks.